src/cargo/util/config/mod.rs

```diff
@@ -483,6 +483,23 @@ impl Config {
         Ok(())
     }
 
+    /// Updates the application's notion of current working directory, both in
+    /// the process environment as well as the configuration data
+    pub fn set_cwd(&mut self, new_cwd: &Path) -> CargoResult<()> {
+        // Update the process-level notion of cwd
+        std::env::set_current_dir(&new_cwd)?;
+        // Update struct members derived from the process-wide cwd
+        // processing occurs
+        self.cwd = new_cwd.to_path_buf();
+        self.home_path = Filesystem::new(homedir(new_cwd).ok_or_else(|| {
```

Comment:
Would it be possible for us to delay the initialization of these?  It'd help catch mistakes of relying on the wrong CWD and it would mean we could call `set_current_dir` out of this function and then just do this initialization step reading from `current_dir`

Reply:
If I understand your question/suggestion correctly, `Config::home_path` and `Config::cwd` could be changed to LazyCell, and then the cli.rs `main()` could just call `set_current_dir()` right after the cli args are fetched.  And `main()` already has a stern-ish warning about accessing its values before it's properly configured near the end of `main()`.

Let me know if that's what you want.  It would affect `Config::new()`'s signature, because it wouldn't be allowed to take `cwd` or `home_path` anymore, and the implementation of `Config::default()` would be much simplified.  At that point, it could even be argued that `Config::new()`, having no callers other than `Cargo::default()` (that I could find) should be eliminated.

Reply:
#11029

Reply:
#11029 got merged. I guess we can move forward. Thanks!